ql/termstructures/yield/fittedbonddiscountcurve.hpp

```diff
@@ -205,6 +206,10 @@ namespace QuantLib {
         ext::shared_ptr<OptimizationMethod> optimizationMethod() const;
         //! open discountFunction to public
         DiscountFactor discount(const Array& x, Time t) const;
+        //! set constraint for the fitting method.
+        void setConstraint(const Constraint& constraint);
```

Comment:
This class doesn't have any other setters, so I think it might be better to pass the `Constraint` instance to the constructor and specify its default value in the constructor as well.

Reply:
Okay, thanks, but I have two objections:
1. Expanding the ctor breaks binary compatiblity.
2. Expanding the ctor for one rarely used feature does not seem practical.

Reply:
I might be wrong, but I don't think QuantLib has ever guaranteed binary compatibility between versions. Either way, you should definitely recompile the library when upgrading to a new version.

There are lots of lesser-used constructor arguments that are defaulted in QuantLib. including the existing arguments to `FittingMethod`, so I don't see anything impractical about defaulting `Contraint` as well.

Reply:
Well, what is the rationale behind cluttering up the constructor with every member variable the object has, at the same time providing default values for most of them, when only a very few of them - if any - are really needed for construction, and most of them are optional? 
What is the name of this pattern?

At the moment, I am sorry, I am not yet convinced that providing the constructor with yet another parameter would be advantageous. 
In fact, I think it would be mostly disadvantageous, in particular due to forcing every single client of the library who inherited from FittingMethod, to change her code, even if she doesn't care at all.

Reply:
> Well, what is the rationale behind cluttering up the constructor with every member variable the object has, at the same time providing default values for most of them, when only a very few of them - if any - are really needed for construction, and most of them are optional?

All of the constructor arguments to `FittingMethod` are optional in the sense that they all have a default value, which will be used to construct the object if the user doesn't pass a value explicitly. But none of the constructor arguments is optional in the sense that all class member variables need to be properly initialized with either a default value or a user-provided value by the time the constructor exits.

> What is the name of this pattern?

Not sure what you are referring to, but the basic principle is both "be consistent" (i.e. with the existing code style), and "minimize statefulness" (i.e. by not exposing a setter).

> In fact, I think it would be mostly disadvantageous, in particular due to forcing every single client of the library who inherited from FittingMethod, to change her code, even if she doesn't care at all.

That is not true, subclasses of `FittingMethod` do not need to change because the default constructor value will be used, as mentioned above.

Reply:
Sorry for disappearing — busy period.

I think @CayOest means that, if someone inherited from `FittingMethod`, they would need to change the constructor of their class in order to use the new constraint feature, while the setter would be inherited.  @sweemer means that user code wouldn't break.  Both statements are true.

This said, our usual rationale is that we tend to avoid setters for two reasons.

One is that they interact with the observer/observable pattern and need special treatment.  For instance, `FittedBondDiscountCurve` inherits from `LazyObject`, meaning that the fit is performed when needed and then cached until the class receives notification that something is changed. With the added setter as it is now, this code wouldn't work as probably intended:
```
FittedBondDiscountCurve curve(...);  // data are stored but calibration is delayed
auto f1 = curve.discount(t);         // this triggers calibration since we need it
curve.setConstraint(...);            // this doesn't signal the curve that it needs recalibration
auto f2 = curve.discount(t);         // this will be the same as f1 
```
To work as expected, `setConstraint` should at least include a call to `update()`.

The second (less probable, but we've been bitten by something like this in the past) is that curves can be shared among any number of instruments, and if some part of the code modifies them in order to do some analysis on one of the instruments, this would affect all other instruments as well, which might not be intended.

As usual, I don't think there's a single obviously correct answer.  Let me know what you think.

Reply:
In addition to the points you raised @lballabio I would like to stand by my points and vote for the defaulted constructor argument approach.

Reply:
I'm in line with @sweemer here. Immutable objects aka "_minimize statefulness_" are one of the best guards against unexpected behavior like @lballabio showed above.

Also we have `RelinkableHandle<T>::linkTo(...)` to change an underlying `TermStructure` consistently when curves are "_shared among any number of instruments_".

Reply:
> Sorry for disappearing — busy period.
> 
> I think @CayOest means that, if someone inherited from `FittingMethod`, they would need to change the constructor of their class in order to use the new constraint feature, while the setter would be inherited. @sweemer means that user code wouldn't break. Both statements are true.
> 
> This said, our usual rationale is that we tend to avoid setters for two reasons.
> 
> One is that they interact with the observer/observable pattern and need special treatment. For instance, `FittedBondDiscountCurve` inherits from `LazyObject`, meaning that the fit is performed when needed and then cached until the class receives notification that something is changed. With the added setter as it is now, this code wouldn't work as probably intended:
> 
> ```
> FittedBondDiscountCurve curve(...);  // data are stored but calibration is delayed
> auto f1 = curve.discount(t);         // this triggers calibration since we need it
> curve.setConstraint(...);            // this doesn't signal the curve that it needs recalibration
> auto f2 = curve.discount(t);         // this will be the same as f1 
> ```
> 
> To work as expected, `setConstraint` should at least include a call to `update()`.
> 
> The second (less probable, but we've been bitten by something like this in the past) is that curves can be shared among any number of instruments, and if some part of the code modifies them in order to do some analysis on one of the instruments, this would affect all other instruments as well, which might not be intended.
> 
> As usual, I don't think there's a single obviously correct answer. Let me know what you think.

Great reply! 
Thanks for the explanation!

I can work with this - not with: We have always done it like this! ;)

Reply:
To be frank: I was just wondering why you would choose such a weird, unnatural pattern.

But thanks to you, I am not convinced.
So you chose it, because you like the Observer pattern, which has nothing, really nothing to do with an algorithmic library.

Reply:
Hi @CayOest,

I think we are not that far away from each other as it might look like:

- I guess we all agree that making `constraint` configurable is worth this PR and we only differ in the way to achieve it.

- I haven't read @sweemer initial comment _This class doesn't have any other setters_ the way you perhaps have: _We have always done it like this!_
For me he was already pointing at the possible impacts by **introducing setters** which makes an immutable object mutable and stateful. @lballabio described these impacts later on in more detail.
That's at least what I immediately thought off when he mentioned **introducing setters**. And keeping objects immutable for me is a [strong pattern to follow](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-immutable) and we need good reasons to make objects mutable.

- _We have always done it like this!_ can also be a good argument, can't it? Being consistent is a value of its own. But of course we need good arguments for it (which we have here IMHO).

Also I agree that the Observer pattern looked weird to me at the beginning. But there are good reasons for it. Together with the `LazyObject` pattern it enables us to update market data aka the `RelinkableHandle<T>::linkTo(...)` for all instruments of a big portfolio and informing only the instruments which are dependent of this market data.

Let us assume we calculate a stress test and have calculated the base scenario already. When we want to evaluate the impact of the portfolio after a 20% downshift of European stock prices we only recalculate the prices of the European stocks in there but can avoid potentially long lasting calculations for bonds, swaptions etc. because we have done them in the base scenario already.

In this sense the QuantLib is not a pure _algorithmic library_ to me but it also solves performance issues which automatically comes with Monte Carlo simulations, fd- or tree-calculations when we consistently follow the patterns it has established and which work quiet well.

I'm not sure if I would introduce it in 2023 again if I would have write a library like this from scratch but the patterns were introduced in 2000 already right from the beginning when there were one-core-processors around and performance was a real issue.

Reply:
Ok, wow, thanks a lot for the extensive comment. I highly appreciate it!
It has been very helpful. 

#1 I'd like to apologize to everybody who would be offended by my ignorant comments! I'm actually very
courteous in person.

#2 [quote] In this sense the QuantLib is not a pure algorithmic library to me but it also solves performance issues which automatically comes with Monte Carlo simulations, fd- or tree-calculations when we consistently follow the patterns it has established and which work quiet well.
[/quote]

Well, I'm not sure if QuantLib solves performance issues. 
I don't think it did. I don't think it was ever intended to. I don't think it does.
But what it was intended to do, it did marvelously.

#3 though: The micro code perspective:
As I gather: most of you are actually Python programmers, right?
Most of the the comments are about immutablity - which is actually not an issue whatsoever for C++ developers. 

It is my believe that it is actually an anti-pattern in C++ to blow up constructors like this.
This is completely different from Python.
But I understand the immutability issues Python guys have. And then I don't.

#4 Regarding the update.
Yes, I will provide the update to the set function.
I will not enhance the blow-up constructor.

Do what you will with it. ;)

Reply:
No, I've been a C++ programmer for a lot longer than I've been using Python — almost 30 years now.  If we ever meet over a beer or two, I can tell you stories of the dark times before the C++98 standard, when every compiler release required a different syntax for templates...

Reply:
Yes, dear Professor Ballabio, I'd definitely enjoy that. :)

I think QuantLib is really a paradigm on how to make a quant lib work. 
Obviously, it doesn't fit every use-case in every scenario.
But for most cases it obviously is more than enough.

However, one question, though: Where did your design decision to overcharge (not overload) the C++ ctors in such a way come from?
I'm sure there has to be an explanation. And then I will probably agree. But at the moment, it is my understanding that it's done because of immutability issues. Which - as a boilerplate - always works in Python discussions. They don't work in C++, though. ^^

As far as GPT-4 is concerned, you are living in London. So it's just a matter of time until we need to call you Sir! :)

Kindest regards!

Reply:
GPT often makes stuff up.  I guess it's getting closer to humans?  Anyway, I live in Milano :)

As for the design: it is actually an attempt to keep objects immutable, for reasons that (it seems to me) have little to do with particular languages; setters make it more difficult to check that you're doing things correctly when objects are shared, as is the case with a discount curve that will be used by several instruments; and also in my view they undermine encapsulation, or at least make it more cumbersome to maintain class invariants.  Also, setters work against thread safety, but that's not a particular concern for QuantLib which is not thread-safe anyway,

All of this can be managed, of course (in C++ as in any other language), but avoiding them simplifies things, and so far we had very few use cases that warranted a setter.  The only one that comes to mind is `Instrument::setPricingEngine`.

This said, if you're making the argument that initialization is cumbersome when a constructor has too many arguments, I agree.  But that's a different issue, I think: we can try to make initialization easier, and still prevent the object from being changed after initialization is complete.  For easier construction, we're using some kind of Fluent Interface in a few places; see for instance `MakeVanillaSwap` used [here](https://github.com/lballabio/QuantLib/blob/master/test-suite/swaption.cpp#L387), or `MakeSchedule`, or other similar classes.

Reply:
> All of this can be managed, of course (in C++ as in any other language), but avoiding them simplifies things, and so far we had very few use cases that warranted a setter. The only one that comes to mind is `Instrument::setPricingEngine`.

Perhaps worth mentioning here also is `RelinkableHandle<T>::linkTo(...)` which sets a new link. And like the implementation of `Instrument::setPricingEngine` it takes care that the changed link is reflected correctly by its observers.

https://github.com/lballabio/QuantLib/blob/56fa1a65165024ce2944229d6ca2cfc0b182da94/ql/handle.hpp#L132-L144